asmcomp/cmmgen.ml

```diff
@@ -1660,11 +1660,48 @@ let make_switch arg cases actions dbg =
       | Cconst_symbol (s, _) -> Csymbol_address s
       | _ -> assert false in
     let const_actions = Array.map to_data_item actions in
-    let table = Compilenv.new_const_symbol () in
-    Cmmgen_state.add_constant table (Const_table (Local,
-        Array.to_list (Array.map (fun act ->
-          const_actions.(act)) cases)));
-    addr_array_ref (Cconst_symbol (table, dbg)) (tag_int arg dbg) dbg
+    let linear_function_result =
+      (* In case the resulting integers are a linear function of the index, we don't
+         emit a table, and just compute the result directly *)
+      let length = Array.length const_actions in
+      if length >= 2
+      then begin
+        match const_actions.(0), const_actions.(1) with
+        | Cint v0, Cint v1 ->
+          let offset = v0 in
+          let slope  = Nativeint.sub v1 v0 in
+          let v = ref (Some (offset, slope)) in
+          for i = 2 to pred length do
+            match const_actions.(i) with
+            | Cint value ->
+              if value <> Nativeint.(add offset (mul (of_int i) slope))
+              then v := None
+            | _ -> v:= None
+          done;
+          !v
+        | _, _ ->
+          None
+      end
+      else None
+    in
+    match linear_function_result with
+    | None ->
+      let table = Compilenv.new_const_symbol () in
+      Cmmgen_state.add_constant table (Const_table (Local,
+          Array.to_list (Array.map (fun act ->
+            const_actions.(act)) cases)));
+      addr_array_ref (Cconst_symbol (table, dbg)) (tag_int arg dbg) dbg
+    | Some (offset, slope) ->
+      let const_natint x dbg =
```

Comment:
I'm not really sure about this part, and the distinction between Cconst_natint and Cconst_int. 
I added this section instead of just using nativeint because add_int and mul_int have special cases that optimizes uses of Cconst_int